src/app/data-model-provider/Provider.h

```diff
@@ -90,18 +86,17 @@ class Provider : public ProviderMetadataTree
     ///     - Validation of readability/writability (also controlled by OperationFlags::kInternal)
     ///     - Validation of timed interaction required (also controlled by OperationFlags::kInternal)
     ///
-    /// Return codes
-    ///   CHIP_IM_GLOBAL_STATUS(code):
-    ///       - error codes that are translatable to specific IM codes
-    ///       - in particular, the following codes are interesting/expected
-    ///         - `UnsupportedWrite` for attempts to write read-only data
-    ///         - `UnsupportedAccess` for ACL failures
-    ///         - `NeedsTimedInteraction` for writes that are not timed however are required to be so
-    virtual CHIP_ERROR WriteAttribute(const WriteAttributeRequest & request, AttributeValueDecoder & decoder) = 0;
+    /// Return notes:
+    ///    - The following codes are interesting/expected to be possible:
+    ///      - `UnsupportedWrite` for attempts to write read-only data
+    ///      - `UnsupportedAccess` for ACL failures
+    ///      - `NeedsTimedInteraction` for writes that are not timed however are required to be so
+    virtual ActionReturnStatus WriteAttribute(const WriteAttributeRequest & request, AttributeValueDecoder & decoder) = 0;
 
     /// `handler` is used to send back the reply.
-    ///    - returning a value other than CHIP_NO_ERROR implies an error reply (error and data are mutually exclusive)
-    virtual CHIP_ERROR Invoke(const InvokeRequest & request, chip::TLV::TLVReader & input_arguments, CommandHandler * handler) = 0;
+    ///    - returning a value other than Success implies an error reply (error and data are mutually exclusive)
+    virtual ActionReturnStatus Invoke(const InvokeRequest & request, chip::TLV::TLVReader & input_arguments,
+                                      CommandHandler * handler) = 0;
```

Comment:
Still need to document what it means to return an error here, right?  Does that indicate an action-level failure such that no replies were or will be encoded via the CommandHandler at all?  This is pretty unclear after these changes.

Reply:
Added documentation here to make it clear that returning anything except pure success (i.e. not CHIP_ERROR or Status::Success) will end up sending a StatusIB without any data.

This is slightly awkward because it means ClusterSpecificStatus is success and prevents data. But it seems to be what the spec demands.